graylog2-server/src/main/java/org/graylog/plugins/map/config/S3GeoIpFileService.java

```diff
@@ -0,0 +1,321 @@
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.time.Instant;
+
+/**
+ * Service for pulling Geo Location Processor ASN and city database files from an S3 bucket and storing them on disk.
+ * The files will initially be downloaded to a temporary location on disk, then they will be validated by the
+ * {@link org.graylog2.rest.resources.system.GeoIpResolverConfigValidator}, and after successful validation they will
+ * be moved to the active location so that the Geo Location Processor can read them. The on-disk directory location
+ * for downloaded files is S3_DOWNLOAD_LOCATION in {@link GeoIpProcessorConfig}. The file names are hardcoded to ensure
+ * that the proper files are always left active.
+ *
+ * This service is called from two places:
+ * - {@link org.graylog2.rest.resources.system.GeoIpResolverConfigValidator} will download new files when the Geo
+ * Location Processor configuration is changed and the new configuration has different S3 objects than the old.
+ * - {@link org.graylog.plugins.map.geoip.GeoIpDbFileChangeMonitorService} will check to see if new files need to be
+ * downloaded each time the service runs based on the lastModified times of the S3 objects.
+ *
+ * This class relies on the {@link DefaultCredentialsProvider} and not any settings that may be configured in the
+ * Graylog AWS plugin configuration. See https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html#credentials-chain
+ * for how to configure your environment so that the default provider retrieves credentials properly.
+ */
+@Singleton
+public class S3GeoIpFileService {
+    private static final Logger LOG = LoggerFactory.getLogger(S3GeoIpFileService.class);
+
+    public static final String S3_BUCKET_PREFIX = "s3://";
+    public static final String ACTIVE_ASN_FILE = "asn-from-s3.mmdb";
+    public static final String ACTIVE_CITY_FILE = "standard_location-from-s3.mmdb";
+    public static final String TEMP_ASN_FILE = "temp-" + ACTIVE_ASN_FILE;
+    public static final String TEMP_CITY_FILE = "temp-" + ACTIVE_CITY_FILE;
+    public static final String NULL_S3_CLIENT_MESSAGE = "Unable to create DefaultCredentialsProvider for the S3 Client. Geo Location Processor S3 file refresh is disabled.";
+
+    private S3Client s3Client;
+    private final Path asnPath;
+    private final Path cityPath;
+    private final Path tempAsnPath;
+    private final Path tempCityPath;
+
+    private Instant asnFileLastModified = Instant.EPOCH;
+    private Instant cityFileLastModified = Instant.EPOCH;
+    private Instant tempAsnFileLastModified = null;
+    private Instant tempCityFileLastModified = null;
+
+    @Inject
+    public S3GeoIpFileService(@Named(GeoIpProcessorConfig.S3_DOWNLOAD_LOCATION) String s3DownloadLocation) {
+        try {
+            final S3ClientBuilder clientBuilder = S3Client.builder();
```

Comment:
I think we should avoid initializing the client in the constructor. This executes AWS SDK code at the time when Guice bindings are initialized at Graylog server startup. Although minimal, it does not seem like the right place. 

Would there be any harm to reinitializing the client each time communication with AWS S3 is needed? Another option is to lazily initialize the client once upon the first API request. `Suppliers.memoize` is used in some other areas of Graylog.

https://github.com/Graylog2/graylog2-server/blob/79d1f6f23fd7ba638ab8f8f4a1296eae500ab453/graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/RestHighLevelClientProvider.java#L64

Reply:
I don't see any harm in initializing it each time. This shouldn't be something that get called so frequently that it is an issue. Working on this now.